### PR TITLE
CPM-503: WIP

### PR DESCRIPTION
**Pull request by pierallard**
wip

### CODE REVIEW

File: src/Akeneo/Pim/Enrichment/Bundle/Command/MigrateToUuidFillProductUuid.php [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+
+    private function fillMissingProductUuidsInsert(): void
+    {
+        // TODO CREATE INDEX toto ON pim_catalog_product (uuid);
+
+        $this->connection->beginTransaction();
+        $productIdsWithoutUuids = $this->connection->fetchFirstColumn(sprintf('SELECT id FROM pim_catalog_product WHERE uuid is NULL LIMIT %d', self::BATCH_SIZE));
+        $values = array_map(fn (string $productId): string => sprintf('(%s, UUID_TO_BIN("%s"), 1, CONCAT(md5(rand()), md5(rand())), "{}", NOW(), NOW())', $productId, Uuid::uuid4()->toString()), $productIdsWithoutUuids);
+
+        $sql = sprintf('
```
Comment (rfauglas) [inline src/Akeneo/Pim/Enrichment/Bundle/Command/MigrateToUuidFillProductUuid.php:86]: Are you sure it works with big catalogs? 
mysql> SHOW VARIABLES LIKE 'max_allowed_packet';
+--------------------+----------+
| Variable_name      | Value    |
+--------------------+----------+
| max_allowed_packet | 67108864 |
+--------------------+----------+
1 row in set (0.02 sec)

**Comment by rfauglas** [comment]
Have you been thinking of upgrading  ramsey/uuid?
https://github.com/akeneo/pim-community-dev/pull/15578
See section on ramsey uuid upgrade risk management:
https://akeneo.atlassian.net/browse/BH-952?focusedCommentId=247452

File: src/Akeneo/Pim/Enrichment/Bundle/Doctrine/ORM/Query/QuantifiedAssociation/GetUuidMappingFromProductIdentifiersQuery.php [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+use Akeneo\Pim\Enrichment\Component\Product\Model\QuantifiedAssociation\UuidMapping;
+use Akeneo\Pim\Enrichment\Component\Product\Query\QuantifiedAssociation\GetIdMappingFromProductIdentifiersQueryInterface;
+use Akeneo\Pim\Enrichment\Component\Product\Query\QuantifiedAssociation\GetUuidMappingFromProductIdentifiersQueryInterface;
+use Doctrine\DBAL\Connection;
+
+/**
+ * @copyright 2022 Akeneo SAS (http://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+class GetUuidMappingFromProductIdentifiersQuery implements GetUuidMappingFromProductIdentifiersQueryInterface
```
Comment (StevenVAIDIE) [inline src/Akeneo/Pim/Enrichment/Bundle/Doctrine/ORM/Query/QuantifiedAssociation/GetUuidMappingFromProductIdentifiersQuery.php:17]: Missing integration test

File: src/Akeneo/Pim/Enrichment/Bundle/Doctrine/ORM/Query/QuantifiedAssociation/GetUuidMappingFromProductIdentifiersQuery.php [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+
+use Akeneo\Pim\Enrichment\Component\Product\Model\QuantifiedAssociation\IdMapping;
+use Akeneo\Pim\Enrichment\Component\Product\Model\QuantifiedAssociation\UuidMapping;
+use Akeneo\Pim\Enrichment\Component\Product\Query\QuantifiedAssociation\GetIdMappingFromProductIdentifiersQueryInterface;
+use Akeneo\Pim\Enrichment\Component\Product\Query\QuantifiedAssociation\GetUuidMappingFromProductIdentifiersQueryInterface;
+use Doctrine\DBAL\Connection;
+
+/**
+ * @copyright 2022 Akeneo SAS (http://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
```
Comment (StevenVAIDIE) [inline src/Akeneo/Pim/Enrichment/Bundle/Doctrine/ORM/Query/QuantifiedAssociation/GetUuidMappingFromProductIdentifiersQuery.php:15]: Please use https, i just used my credit card in those two site and I have been hacked
```suggestion
 * @copyright 2022 Akeneo SAS (https://www.akeneo.com)
 * @license   https://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
 ```

File: src/Akeneo/Pim/Enrichment/Bundle/Doctrine/ORM/Query/QuantifiedAssociation/GetUuidMappingFromProductIdentifiersQuery.php [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+ */
+class GetUuidMappingFromProductIdentifiersQuery implements GetUuidMappingFromProductIdentifiersQueryInterface
+{
+    /** @var Connection */
+    private $connection;
+
+    public function __construct(Connection $connection)
+    {
+        $this->connection = $connection;
+    }
```
Comment (StevenVAIDIE) [inline src/Akeneo/Pim/Enrichment/Bundle/Doctrine/ORM/Query/QuantifiedAssociation/GetUuidMappingFromProductIdentifiersQuery.php:25]:
```suggestion
    public function __construct(private Connection $connection)
    {
    }
```

File: src/Akeneo/Pim/Enrichment/Component/Product/Model/QuantifiedAssociation/UuidMapping.php [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+ * @copyright 2022 Akeneo SAS (http://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+final class UuidMapping
+{
+    /** @var array  */
+    private $uuidsToIdentifiers;
+
+    /** @var array  */
+    private $identifiersToUuids;
```
Comment (StevenVAIDIE) [inline src/Akeneo/Pim/Enrichment/Component/Product/Model/QuantifiedAssociation/UuidMapping.php:21]:
```suggestion
    /** @var array<string, string> */
    private array $uuidsToIdentifiers;

    /** @var array<string, string> */
    private array $identifiersToUuids;
```

File: tests/back/Pim/Enrichment/Specification/Component/Product/Model/QuantifiedAssociation/UuidMappingSpec.php [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+
+        $this->getIdentifier(Uuid::fromString($uuid))->shouldReturn($identifier);
+        $this->getUuid($identifier)->equals($uuid)->shouldBe(true);
+        $this->hasUuid($identifier)->shouldReturn(true);
+        $this->hasIdentifier($uuid)->shouldReturn(true);
+        $this->hasUuid('nice')->shouldReturn(false);
+        $this->hasIdentifier(Uuid::fromString('52254bba-a2c8-40bb-abe1-195e3970bd93'))->shouldReturn(false);
+    }
+
+    public function it_throws_if_the_product_id_is_not_an_integer()
```
Comment (StevenVAIDIE) [inline tests/back/Pim/Enrichment/Specification/Component/Product/Model/QuantifiedAssociation/UuidMappingSpec.php:43]:
```suggestion
    public function it_throws_if_the_product_id_is_not_an_uuid
```

**Review by StevenVAIDIE — APPROVED** [review]

File: src/Akeneo/Pim/Enrichment/Bundle/Command/MigrateToUuidFillProductUuid.php (hunk 1 above)
Comment (pierallard) [reply src/Akeneo/Pim/Enrichment/Bundle/Command/MigrateToUuidFillProductUuid.php:86]: It's done by batches of 1000. This should be OK.

**Comment by pierallard** [comment]
> Have you been thinking of upgrading ramsey/uuid? #15578 See section on ramsey uuid upgrade risk management: https://akeneo.atlassian.net/browse/BH-952?focusedCommentId=247452

This is clearly out of the scope of this particular PR

File: src/Akeneo/Tool/Bundle/VersioningBundle/Doctrine/ORM/VersionRepository.php [hunk 7]
```diff
@@ -33,6 +33,10 @@ public function getLogEntries($resourceName, $resourceId)
         );
     }
 
+    public function getProductEntries($resourceId)
```
Comment (pierallard) [inline src/Akeneo/Tool/Bundle/VersioningBundle/Doctrine/ORM/VersionRepository.php:36]: remove

**Pull request merged**